rsconnect/api.py

```diff
@@ -268,6 +268,312 @@ def output_task_log(task_status, last_status, log_callback):
         return new_last_status
 
 
+class RSConnectExecutor:
+    def __init__(self, *args, **kwargs) -> None:
```

Comment:
To make this API more usable and support tooling, we need named arguments with types here. Rather than having lots of arguments as the functions in actions.py currently do, we can group them as we have done in main.py with the `@server_args` and `@content_args` decorators. If those decorators each consumed multiple CLI arguments and bundled them into an object (perhaps a dataclass, which is usable as far back as Python 3.6), then the CLI functions would accept a `server` argument and/or a `content` argument which could be passed through to this constructor.

Reply:
This is an interesting design decision to think through. The design so far is intended to push the complexity down. That comes with some advantages and disadvantages. The advantages being that the user can provide a grab bag of keyword arguments and we'll setup the Connect backend target correctly, or warn the user that something is missing. 

Let's say that we switched over to use another `RSConnectServer`, but also have to handle backwards compatibility; instead of having the user (or the `@server_args` in this case) worry about constructing the correct version of `RSConnectServer`, we just seamlessly handle it within the Executor.

This is probably more evident in how we are refactoring to encapsulate `RSConnect(HTTPServer)`. If we need to switch `RSConnect(HTTPServer)` to another backend target like connect-cloud, the user can just pass another keyword argument instead of passing us a fully constructed http server.